File: 102.binary-tree-level-order-traversal.js
/*
 * @lc app=leetcode id=102 lang=javascript
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function (root) {
  const hashTable = {};

  const dfs = (node, level) => {
    if (!node) return;

    if (!hashTable[level]) hashTable[level] = [];
    hashTable[level].push(node.val);

    if (node.left) dfs(node.left, level + 1);
    if (node.right) dfs(node.right, level + 1);
  };

  if (!root) return [];
  hashTable[1] = [root.val];

  if (root.left) dfs(root.left, 2);
  if (root.right) dfs(root.right, 2);

  return Object.values(hashTable);
};
// @lc code=end

File: 102.binary-tree-level-order-traversal.java
/*
 * @lc app=leetcode id=102 lang=java
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        Map<Integer, List<Integer>> hashTable = new HashMap<>();

        if (root == null)
            return new ArrayList<>();

        hashTable.put(1, new ArrayList<>(List.of(root.val)));

        dfs(root.left, 2, hashTable);
        dfs(root.right, 2, hashTable);

        return new ArrayList<>(hashTable.values());

    }

    public void dfs(TreeNode node, int level, Map<Integer, List<Integer>> hashTable) {
        if (node == null)
            return;

        if (!hashTable.containsKey(level)) {
            hashTable.put(level, new ArrayList<>());
        }
        hashTable.get(level).add(node.val);

        dfs(node.left, level + 1, hashTable);
        dfs(node.right, level + 1, hashTable);
    }

}
// @lc code=end

